File: packages/maencof/skills/craft-skill/scripts/mode_detector.py
#!/usr/bin/env python3
"""
Mode Detector - Determines CREATE/REFACTOR/IMPROVE/FIX from context

Usage:
    mode_detector.py --request "user request text" [--skill-path <path>]

Output:
    JSON: {"mode": "CREATE", "confidence": 0.95, "reasons": [...]}
"""

import argparse
import json
import re
from pathlib import Path
from typing import Dict, List, Optional


class Mode:
    CREATE = "CREATE"
    REFACTOR = "REFACTOR"
    IMPROVE = "IMPROVE"
    FIX = "FIX"


def detect_mode(request: str, skill_path: Optional[Path] = None) -> Dict:
    """
    Detect operating mode from user request and context.

    Returns:
        {
            "mode": str,
            "confidence": float (0-1),
            "reasons": List[str],
            "alternatives": List[str],
            "existing_skill": bool
        }
    """
    request_lower = request.lower()
    existing_skill = skill_path and skill_path.exists() and (skill_path / "SKILL.md").exists()

    # Keyword analysis
    create_keywords = ["create", "new skill", "build", "initialize", "start"]
    refactor_keywords = ["refactor", "restructure", "reorganize", "clean up structure"]
    improve_keywords = ["improve", "enhance", "add feature", "extend", "upgrade"]
    fix_keywords = ["fix", "bug", "issue", "error", "broken", "debug", "resolve"]

    scores = {
        Mode.CREATE: sum(1 for kw in create_keywords if kw in request_lower),
        Mode.REFACTOR: sum(1 for kw in refactor_keywords if kw in request_lower),
        Mode.IMPROVE: sum(1 for kw in improve_keywords if kw in request_lower),
        Mode.FIX: sum(1 for kw in fix_keywords if kw in request_lower),
    }

    # Context adjustment
    if not existing_skill:
        scores[Mode.CREATE] += 5  # Strong bias for CREATE if no existing skill
        scores[Mode.REFACTOR] = 0
        scores[Mode.IMPROVE] = 0
        scores[Mode.FIX] = 0

    # Determine mode
    max_score = max(scores.values())
    if max_score == 0:
        # No clear keywords - default based on existence
        mode = Mode.CREATE if not existing_skill else Mode.IMPROVE
        confidence = 0.5
        reasons = ["No specific keywords detected, using default"]
    else:
        mode = max(scores, key=scores.get)
        confidence = min(max_score / 5.0, 1.0)  # Normalize to 0-1
        reasons = [f"Detected keywords for {mode}"]

    if not existing_skill and mode != Mode.CREATE:
        reasons.append("No existing skill found, defaulting to CREATE")
        mode = Mode.CREATE

    if existing_skill:
        reasons.append("Existing skill detected")

    # Find alternatives
    alternatives = [m for m, s in sorted(scores.items(), key=lambda x: x[1], reverse=True)
                    if m != mode and s > 0]

    return {
        "mode": mode,
        "confidence": round(confidence, 2),
        "reasons": reasons,
        "alternatives": alternatives[:2],  # Top 2 alternatives
        "existing_skill": existing_skill
    }


def main():
    parser = argparse.ArgumentParser(description="Detect skill creation mode")
    parser.add_argument("--request", required=True, help="User request text")
    parser.add_argument("--skill-path", help="Path to existing skill directory")
    args = parser.parse_args()

    skill_path = Path(args.skill_path) if args.skill_path else None
    result = detect_mode(args.request, skill_path)

    print(json.dumps(result, indent=2))


if __name__ == "__main__":
    main()

File: packages/maencof/skills/maencof-craft-skill/scripts/mode_detector.mjs
#!/usr/bin/env node
/**
 * Mode Detector - Determines CREATE/REFACTOR/IMPROVE/FIX from context
 *
 * Usage:
 *     node mode_detector.mjs --request "user request text" [--skill-path <path>]
 *
 * Output:
 *     JSON: {"mode": "CREATE", "confidence": 0.95, "reasons": [...]}
 */

import { parseArgs } from 'node:util';
import { existsSync } from 'node:fs';
import path from 'node:path';

const Mode = Object.freeze({
  CREATE: 'CREATE',
  REFACTOR: 'REFACTOR',
  IMPROVE: 'IMPROVE',
  FIX: 'FIX',
});

// Weighted keyword definitions: [keyword, weight]
const KEYWORD_WEIGHTS = {
  [Mode.CREATE]: [
    ['create', 3], ['new skill', 4], ['build', 2], ['initialize', 3],
    ['start', 1], ['from scratch', 4], ['make a skill', 3], ['generate', 2],
  ],
  [Mode.REFACTOR]: [
    ['refactor', 4], ['restructure', 4], ['reorganize', 3], ['clean up structure', 3],
    ['split', 2], ['extract', 2], ['simplify structure', 3], ['slim down', 2],
  ],
  [Mode.IMPROVE]: [
    ['improve', 3], ['enhance', 3], ['add feature', 4], ['extend', 2],
    ['upgrade', 2], ['add support', 3], ['integrate', 2], ['expand', 2],
  ],
  [Mode.FIX]: [
    ['fix', 4], ['bug', 3], ['issue', 2], ['error', 3], ['broken', 4],
    ['debug', 3], ['resolve', 2], ['not working', 4], ['fails', 3],
  ],
};

function calculateModeScores(request) {
  const requestLower = request.toLowerCase();
  const scores = { [Mode.CREATE]: 0, [Mode.REFACTOR]: 0, [Mode.IMPROVE]: 0, [Mode.FIX]: 0 };

  for (const [mode, keywords] of Object.entries(KEYWORD_WEIGHTS)) {
    for (const [keyword, weight] of keywords) {
      if (requestLower.includes(keyword)) {
        scores[mode] += weight;
      }
    }
  }

  return scores;
}

function detectMode(request, skillPath) {
  const existingSkill = Boolean(
    skillPath && existsSync(skillPath) && existsSync(path.join(skillPath, 'SKILL.md'))
  );
  const scores = calculateModeScores(request);

  if (!existingSkill) {
    scores[Mode.CREATE] += 3;
  } else {
    scores[Mode.IMPROVE] += 1;
  }

  const maxScore = Math.max(...Object.values(scores));
  const reasons = [];
  let mode, confidence;

  if (maxScore === 0) {
    mode = existingSkill ? Mode.IMPROVE : Mode.CREATE;
    confidence = 0.4;
    reasons.push('No specific keywords detected, using context-based default');
  } else {
    mode = Object.entries(scores).reduce((a, b) => (b[1] > a[1] ? b : a))[0];
    confidence = Math.max(Math.min(maxScore / 8.0, 1.0), 0.3);

    const matchedKeywords = [];
    for (const [keyword] of (KEYWORD_WEIGHTS[mode] || [])) {
      if (request.toLowerCase().includes(keyword)) {
        matchedKeywords.push(keyword);
      }
    }
    if (matchedKeywords.length > 0) {
      reasons.push(`Matched keywords: ${matchedKeywords.join(', ')}`);
    }
  }

  if (existingSkill) {
    reasons.push('Existing skill detected at provided path');
  } else {
    reasons.push('No existing skill found' + (mode === Mode.CREATE ? ' — defaulting to CREATE' : ''));
  }

  const alternatives = Object.entries(scores)
    .sort((a, b) => b[1] - a[1])
    .filter(([m, s]) => m !== mode && s > 0)
    .map(([m]) => m)
    .slice(0, 2);

  return {
    mode,
    confidence: Math.round(confidence * 100) / 100,
    reasons,
    alternatives,
    existing_skill: existingSkill,
  };
}

const { values } = parseArgs({
  options: {
    request: { type: 'string' },
    'skill-path': { type: 'string' },
  },
  strict: true,
});

if (!values.request) {
  console.error('Usage: node mode_detector.mjs --request "user request text" [--skill-path <path>]');
  process.exit(1);
}

const result = detectMode(values.request, values['skill-path'] || null);
console.log(JSON.stringify(result, null, 2));
